cub/test/catch2_test_device_scan_by_key.cu

```diff
@@ -0,0 +1,453 @@
+                                 init_t{},
+                                 num_items,
+                                 eq_op_t{});
+
+    // Verify result
+    REQUIRE(expected_result == out_values);
+
+    // Run test in-place
+    CUB_IF_CONSTEXPR(std::is_same<value_t, output_t>::value)
+    {
+      // Copy input values to memory allocated for output values, to ensure in_values are
+      // unchanged for a (potentially) subsequent test that uses in_values as input
+      out_values            = in_values;
+      auto values_in_out_it = thrust::raw_pointer_cast(out_values.data());
+      device_exclusive_scan_by_key(d_keys_it,
+                                   unwrap_it(values_in_out_it),
+                                   unwrap_it(values_in_out_it),
+                                   scan_op,
+                                   init_t{},
+                                   num_items,
+                                   eq_op_t{});
+
+      // Verify result
+      REQUIRE(expected_result == out_values);
+    }
+  }
+}
+
+#if TEST_TYPES == 0
+using key_alias_type_list = c2h::type_list<std::uint8_t>;
+#elif TEST_TYPES == 1
+using key_alias_type_list = c2h::type_list<std::int32_t>;
+#elif TEST_TYPES == 2
+using key_alias_type_list = c2h::type_list<float>;
+#elif TEST_TYPES == 3
+using key_alias_type_list = c2h::type_list<custom_t>;
+#endif
+
+CUB_TEST("Device scan works when memory for keys and results alias one another",
+         "[by_key][scan][device]",
+         key_alias_type_list)
+{
+  using key_t    = typename c2h::get<0, TestType>;
+  using value_t  = key_t;
+  using output_t = key_t;
+  using offset_t = std::uint32_t;
+
+  constexpr offset_t min_items = 1;
+  constexpr offset_t max_items = 1000000;
+
+  // Generate the input sizes to test for
+  const offset_t num_items = GENERATE_COPY(take(2, random(min_items, max_items)),
+                                           values({
+                                             min_items,
+                                             max_items,
+                                           }));
+  INFO("Test num_items: " << num_items);
+
+  // Range of segment sizes to generate (a segment is a series of consecutive equal keys)
+  const std::tuple<offset_t, offset_t> seg_size_range =
+    GENERATE_COPY(table<offset_t, offset_t>({{1, 1}, {1, num_items}, {num_items, num_items}}));
+  INFO("Test seg_size_range: [" << std::get<0>(seg_size_range) << ", "
+                                << std::get<1>(seg_size_range) << "]");
+
+  // Generate input segments
+  thrust::device_vector<offset_t> segment_offsets =
+    c2h::gen_uniform_offsets<offset_t>(CUB_SEED(1),
+                                       num_items,
+                                       std::get<0>(seg_size_range),
+                                       std::get<1>(seg_size_range));
+
+  // Get array of keys from segment offsets
+  const offset_t num_segments = static_cast<offset_t>(segment_offsets.size() - 1);
```

Comment:
important: unused, let's remove to make msvc happy

Reply:
Good catch, thanks 👍